File: code/Java/Chapter8/8-6/MergeSort.java
import java.util.Arrays;

public class MergeSort {
    private static final int INSERTION_SORT_THRESHOLD = 10;

    public static void main(String[] args) {
        int[] arr = {39, 27, 43, 3, 9, 82, 10};
        mergeSort(arr);
        System.out.println(Arrays.toString(arr));
    }

    /**
     * Binary Insertion Sort
     */
    private static int[] insertionSort(int[] arr) {
        for (int i = 1; i < arr.length; i++) {
            int temp = arr[i];
            int left = 0;
            int right = i - 1;

            while (left <= right) {
                int mid = (left + right) / 2;
                if (temp < arr[mid]) {
                    right = mid - 1;
                } else {
                    left = mid + 1;
                }
            }

            for (int j = i - 1; j >= left; j--) {
                arr[j + 1] = arr[j];
            }
            arr[left] = temp;
        }

        return arr;
    }

    /**
     * Merge Sort (original v1.0)
     */
    public static int[] mergeSort(int[] arr) {
        if (arr.length <= 1) {
            return arr;
        }

        int mid = arr.length / 2;
        int[] leftHalf = Arrays.copyOfRange(arr, 0, mid);
        int[] rightHalf = Arrays.copyOfRange(arr, mid, arr.length);

        mergeSort(leftHalf);
        mergeSort(rightHalf);

        int i = 0;
        int j = 0;
        int k = 0;

        while (i < leftHalf.length && j < rightHalf.length) {
            if (leftHalf[i] < rightHalf[j]) {
                arr[k] = leftHalf[i];
                i++;
            } else {
                arr[k] = rightHalf[j];
                j++;
            }
            k++;
        }

        while (i < leftHalf.length) {
            arr[k] = leftHalf[i];
            i++;
            k++;
        }

        while (j < rightHalf.length) {
            arr[k] = rightHalf[j];
            j++;
            k++;
        }

        return arr;
    }

    /**
     * Merge Sort (optimized v2.0)
     */
    // public static int[] mergeSort(int[] arr) {
    //     if (arr.length <= INSERTION_SORT_THRESHOLD) {
    //         return insertionSort(arr);
    //     }

    //     int mid = arr.length / 2;
    //     int[] leftHalf = Arrays.copyOfRange(arr, 0, mid);
    //     int[] rightHalf = Arrays.copyOfRange(arr, mid, arr.length);

    //     mergeSort(leftHalf);
    //     mergeSort(rightHalf);

    //     int i = 0;
    //     int j = 0;
    //     int k = 0;

    //     while (i < leftHalf.length && j < rightHalf.length) {
    //         if (leftHalf[i] < rightHalf[j]) {
    //             arr[k] = leftHalf[i];
    //             i++;
    //         } else {
    //             arr[k] = rightHalf[j];
    //             j++;
    //         }
    //         k++;
    //     }

    //     while (i < leftHalf.length) {
    //         arr[k] = leftHalf[i];
    //         i++;
    //         k++;
    //     }

    //     while (j < rightHalf.length) {
    //         arr[k] = rightHalf[j];
    //         j++;
    //         k++;
    //     }

    //     return arr;
    // }
}

File: code/C++/Chapter8/8-6/merge_sort.cpp
#include <iostream>

using namespace std;

const int INSERTION_SORT_THRESHOLD = 10;

/**
 * Binary Insertion Sort
 */
template <typename T>
T *insertion_sort(T *arr, int n) {
    for (int i = 1; i < n; i++) {
        T temp = arr[i];
        int left = 0;
        int right = i - 1;

        while (left <= right) {
            int mid = (left + right) / 2;
            if (temp < arr[mid]) {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }

        for (int j = i - 1; j >= left; j--) {
            arr[j + 1] = arr[j];
        }
        arr[left] = temp;
    }

    return arr;
}

/**
 * Merge Sort (original v1.0)
 */
template <typename T>
T *merge_sort(T *arr, int n) {
    if (n <= 1) {
        return arr;
    }

    int mid = n / 2;
    T left_half[mid];
    T right_half[n - mid];
    memcpy(left_half, arr, mid * sizeof(T));
    memcpy(right_half, arr + mid, (n - mid) * sizeof(T));

    merge_sort(left_half, mid);
    merge_sort(right_half, n - mid);

    int i = 0;
    int j = 0;
    int k = 0;

    while (i < mid && j < n - mid) {
        if (left_half[i] < right_half[j]) {
            arr[k] = left_half[i];
            i++;
        } else {
            arr[k] = right_half[j];
            j++;
        }
        k++;
    }

    while (i < mid) {
        arr[k] = left_half[i];
        i++;
        k++;
    }

    while (j < n - mid) {
        arr[k] = right_half[j];
        j++;
        k++;
    }

    return arr;
}

/**
 * Merge Sort (optimized v2.0)
 */
// template <typename T>
// T *merge_sort(T *arr, int n) {
//     if (n <= INSERTION_SORT_THRESHOLD) {
//         return insertion_sort(arr, n);
//     }

//     int mid = n / 2;
//     T left_half[mid];
//     T right_half[n - mid];
//     memcpy(left_half, arr, mid * sizeof(T));
//     memcpy(right_half, arr + mid, (n - mid) * sizeof(T));

//     merge_sort(left_half, mid);
//     merge_sort(right_half, n - mid);

//     int i = 0;
//     int j = 0;
//     int k = 0;

//     while (i < mid && j < n - mid) {
//         if (left_half[i] < right_half[j]) {
//             arr[k] = left_half[i];
//             i++;
//         } else {
//             arr[k] = right_half[j];
//             j++;
//         }
//         k++;
//     }

//     while (i < mid) {
//         arr[k] = left_half[i];
//         i++;
//         k++;
//     }

//     while (j < n - mid) {
//         arr[k] = right_half[j];
//         j++;
//         k++;
//     }

//     return arr;
// }

int main() {
    int arr[] = {39, 27, 43, 3, 9, 82, 10};
    int n = sizeof(arr) / sizeof(arr[0]);

    merge_sort(arr, n);
    for (int i = 0; i < n; i++) {
        cout << arr[i] << " ";
    }
    cout << endl;

    return 0;
}